File: windows/libbibledit/libsrc/navigation/passage.php
<?php
/*
Copyright (©) 2003-2014 Teus Benschop.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/


require_once ("bootstrap/bootstrap.php");


class Navigation_Passage
{


  public static function getContainer ()
  {
    return '<div id="versepickerwrapper"></div>';
  }


  public static function getNavigator ($bible)
  {
    $ipc_focus = Ipc_Focus::getInstance();
    $database_books = Database_Books::getInstance ();
    $database_bibles = Database_Bibles::getInstance ();
    $database_navigation = Database_Navigation::getInstance ();
    $session_logic = Session_Logic::getInstance ();

    $user = $session_logic->currentUser ();

    $fragments = array ();

    // Links to go back and forward are grayed out or active depending on available passages to go to.
    $fragments [] = '<div id="backforward">';
    if ($database_navigation->previousExists ($user)) {
      $fragments [] = '<a id="navigateback" href="navigateback" title="' . Locale_Translate::_("Back") . '">↶</a>';
    } else {
      $fragments [] = '<span class="grayedout">↶</span>';
    }
    $fragments [] = "";
    if ($database_navigation->nextExists ($user)) {
      $fragments [] = '<a id="navigateforward" href="navigateforward" title="' . Locale_Translate::_("Forward") . '">↷</a>';
    } else {
      $fragments [] = '<span class="grayedout">↷</span>';
    }
    $fragments [] = '</div>';

    $book = $ipc_focus->getBook ();

    // The book should exist in the Bible.
    if ($bible != "") {
      $books = $database_bibles->getBooks ($bible);
      if (!in_array ($book, $books)) {
        if (count ($books) > 0) $book = $books [0];
        else $book = 0;
      }
    }

    $bookName = $database_books->getEnglishFromId ($book);

    $chapter = $ipc_focus->getChapter ();

    // The chapter should exist in the book.
    if ($bible != "") {
      $chapters = $database_bibles->getChapters ($bible, $book);
      if (!in_array ($chapter, $chapters)) {
        if (count ($chapters) > 0) $chapter = $chapters [0];
        else $chapter = 1;
      }
    }

    $verse = $ipc_focus->getVerse ();

    $fragments [] = '<div id="versepicker">';

    $fragments [] = '<input name="selectpassage" id="selectpassage" type="text" value="' . "$bookName $chapter:$verse" . '" size="14" />';
    $fragments [] = '<input name="submitpassage" id="submitpassage" type="submit" value="' . Locale_Translate::_("Go") . '" />';

    $fragments [] = '<div id="handpicker">';

    $fragments [] = '<select id="booklist" size=15>';
    $fragments [] = self::getBooksFragment ($bible);
    $fragments [] = '</select>';

    $fragments [] = '<select id="chapterlist" size=15>';
    $fragments [] = self::getChaptersFragment ($bible, $book, $chapter);
    $fragments [] = '</select>';

    $fragments [] = '<select id="verselist" size=15 multiple>';
    $fragments [] = self::getVersesFragment ($bible, $book, $chapter, $verse);

    $fragments [] = '</select>';

    $fragments [] = '</div>';

    $fragments [] = '</div>';

    // The result.
    return implode ("\n", $fragments);
  }


  public static function getBooksFragment ($bible)
  {
    $database_bibles = Database_Bibles::getInstance ();
    $database_books = Database_Books::getInstance ();
    $ipc_focus = Ipc_Focus::getInstance();
    $activeBook = $ipc_focus->getBook ();
    // Take standard books in case of no Bible.
    if ($bible == "") {
      $books = $database_books->getIDs ();
    } else {
      $books = Filter_Books::getOrdered ($bible);
    }
    $html = "";
    foreach ($books as $book) {
      $bookName = $database_books->getEnglishFromId ($book);
      $selected = "";
      if ($book == $activeBook) $selected = " selected";
      $html .= "<option$selected>" . $bookName . "</option>";
    }
    return $html;
  }


  public static function getChaptersFragment ($bible, $book, $chapter)
  {
    if ($bible == "") {
      $database_versifications = Database_Versifications::getInstance ();
      $chapters = $database_versifications->getChapters ("English", $book, true);
    } else {
      $database_bibles = Database_Bibles::getInstance ();
      $chapters = $database_bibles->getChapters ($bible, $book);
    }
    $html = "";
    foreach ($chapters as $ch) {
      $selected = "";
      if ($ch == $chapter) $selected = " selected";
      $html .= "<option$selected>" . $ch . "</option>";;
    }
    return $html;
  }


  public static function getVersesFragment ($bible, $book, $chapter, $verse)
  {
    $database_bibles = Database_Bibles::getInstance ();
    if ($bible == "") {
      $database_versifications = Database_Versifications::getInstance ();
      $verses = $database_versifications->getVerses ("English", $book, $chapter);
    } else {
      $verses = Filter_Usfm::getVerseNumbers ($database_bibles->getChapter ($bible, $book, $chapter));
    }
    $html = "";
    foreach ($verses as $vs) {
      $selected = "";
      if ($verse == $vs) $selected = " selected";
      $html .= "<option$selected>" . $vs . "</option>";;
    }
    return $html;
  }


  public static function code ($bible, $header = false)
  {
    $code = '';
    if ($header) {
      $code .= '<script type="text/javascript" src="../navigation/passage.js"></script>';
      $code .= "\n";
    }
    $code .= '<script type="text/javascript">';
    $code .= "\n";
    $code .= 'navigationBible = "' . $bible . '";';
    $code .= "\n";
    $code .= '</script>';
    $code .= "\n";
    return $code;
  }


  public static function setBookChapterVerse ($book, $chapter, $verse)
  {
    $book = Filter_Numeric::integer_in_string ($book);
    $chapter = Filter_Numeric::integer_in_string ($chapter);
    $verse = Filter_Numeric::integer_in_string ($verse);
    $ipc_focus = Ipc_Focus::getInstance();
    $ipc_focus->set ($book, $chapter, $verse);
    Navigation_Passage::recordHistory ($book, $chapter, $verse);
  }


  public static function setPassage ($bible, $passage)
  {
    $ipc_focus = Ipc_Focus::getInstance();
    $currentBook = $ipc_focus->getBook ();
    $currentChapter = $ipc_focus->getChapter ();
    $currentVerse = $ipc_focus->getVerse ();
    $passage = trim ($passage);
    if (($passage == "") || ($passage == "+")) {
      $passage = Navigation_Passage::getNextVerse ($bible, $currentBook, $currentChapter, $currentVerse);
    } else if ($passage == "-") {
      $passage = Navigation_Passage::getPreviousVerse ($bible, $currentBook, $currentChapter, $currentVerse);
    } else {
      $passage = Filter_Books::interpretPassage (array ($currentBook, $currentChapter, $currentVerse), $passage);
    }
    if ($passage[0] != 0) {
      $ipc_focus->set ($passage [0], $passage [1], $passage [2]);
      Navigation_Passage::recordHistory ($passage [0], $passage [1], $passage [2]);

    }
  }


  private static function getNextVerse ($bible, $book, $chapter, $verse)
  {
    $verse++;
    if ($bible != "") {
      $database_bibles = Database_Bibles::getInstance ();
      $verses = Filter_Usfm::getVerseNumbers ($database_bibles->getChapter ($bible, $book, $chapter));
      if (!in_array ($verse, $verses)) {
        $verse = array_pop ($verses);
      }
    }
    $passage = array ($book, $chapter, $verse);
    return $passage;
  }


  private static function getPreviousVerse ($bible, $book, $chapter, $verse)
  {
    $verse--;
    if ($bible != "") {
      $database_bibles = Database_Bibles::getInstance ();
      $verses = Filter_Usfm::getVerseNumbers ($database_bibles->getChapter ($bible, $book, $chapter));
      if (!in_array ($verse, $verses)) {
        $verse = array_shift ($verses);
      }
    }
    $passage = array ($book, $chapter, $verse);
    return $passage;
  }


  public static function gotoNextVerse ($bible)
  {
    $ipc_focus = Ipc_Focus::getInstance();
    $currentBook = $ipc_focus->getBook ();
    $currentChapter = $ipc_focus->getChapter ();
    $currentVerse = $ipc_focus->getVerse ();
    $passage = Navigation_Passage::getNextVerse ($bible, $currentBook, $currentChapter, $currentVerse);
    if ($passage[0] != 0) {
      $ipc_focus->set ($passage [0], $passage [1], $passage [2]);
      Navigation_Passage::recordHistory ($passage [0], $passage [1], $passage [2]);
    }
  }


  public static function gotoPreviousVerse ($bible)
  {
    $ipc_focus = Ipc_Focus::getInstance();
    $currentBook = $ipc_focus->getBook ();
    $currentChapter = $ipc_focus->getChapter ();
    $currentVerse = $ipc_focus->getVerse ();
    $passage = Navigation_Passage::getPreviousVerse ($bible, $currentBook, $currentChapter, $currentVerse);
    if ($passage[0] != 0) {
      $ipc_focus->set ($passage [0], $passage [1], $passage [2]);
      Navigation_Passage::recordHistory ($passage [0], $passage [1], $passage [2]);
    }
  }


  public static function recordHistory ($book, $chapter, $verse)
  {
    $session_logic = Session_Logic::getInstance ();
    $user = $session_logic->currentUser ();
    $database_navigation = Database_Navigation::getInstance ();
    $database_navigation->record (time (), $user, $book, $chapter, $verse);
  }


  public static function goBack ()
  {
    $database_navigation = Database_Navigation::getInstance ();
    $session_logic = Session_Logic::getInstance ();
    $user = $session_logic->currentUser ();
    $passage = $database_navigation->getPrevious ($user);
    if ($passage) {
      $ipc_focus = Ipc_Focus::getInstance();
      $ipc_focus->set ($passage [0], $passage [1], $passage [2]);
    }
  }


  public static function goForward ()
  {
    $database_navigation = Database_Navigation::getInstance ();
    $session_logic = Session_Logic::getInstance ();
    $user = $session_logic->currentUser ();
    $passage = $database_navigation->getNext ($user);
    if ($passage) {
      $ipc_focus = Ipc_Focus::getInstance();
      $ipc_focus->set ($passage [0], $passage [1], $passage [2]);
    }
  }


}


?>

File: lib/navigation/passage.cpp
/*
 Copyright (©) 2003-2015 Teus Benschop.
 
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */


#include <navigation/passage.h>
#include <filter/string.h>
#include <filter/usfm.h>
#include <database/versifications.h>
#include <database/navigation.h>
#include <database/books.h>
#include <webserver/request.h>
#include <locale/translate.h>
#include <ipc/focus.h>


string Navigation_Passage::getContainer ()
{
  return "<div id=\"versepickerwrapper\"></div>";
}


string Navigation_Passage::getNavigator (void * webserver_request, string bible) // Todo
{
  Webserver_Request * request = (Webserver_Request *) webserver_request;

  Database_Navigation database_navigation = Database_Navigation ();
  
  string user = request->session_logic()->currentUser ();
  
  vector <string> fragments;
  
  // Links to go back and forward are grayed out or active depending on available passages to go to.
  fragments.push_back ("<div id=\"backforward\">");
  if (database_navigation.previousExists (user)) {
    fragments.push_back ("<a id=\"navigateback\" href=\"navigateback\" title=\"" + gettext("Back") + "\">↶</a>");
  } else {
    fragments.push_back ("<span class=\"grayedout\">↶</span>");
  }
  fragments.push_back ("");
  if (database_navigation.nextExists (user)) {
    fragments.push_back ("<a id=\"navigateforward\" href=\"navigateforward\" title=\"" + gettext("Forward") + "\">↷</a>");
  } else {
    fragments.push_back ("<span class=\"grayedout\">↷</span>");
  }
  fragments.push_back ("</div>");
  
  int book = Ipc_Focus::getBook (request);
  
  // The book should exist in the Bible.
  if (bible != "") {
    vector <int> books = request->database_bibles()->getBooks (bible);
    if (find (books.begin(), books.end(), book) == books.end()) {
      if (!books.empty ()) book = books [0];
      else book = 0;
    }
  }
  
  string bookName = Database_Books::getEnglishFromId (book);
  
  int chapter = Ipc_Focus::getChapter (request);
  
  // The chapter should exist in the book.
  if (bible != "") {
    vector <int> chapters = request->database_bibles()->getChapters (bible, book);
    if (find (chapters.begin(), chapters.end(), chapter) == chapters.end()) {
      if (!chapters.empty()) chapter = chapters [0];
      else chapter = 1;
    }
  }
  
  int verse = Ipc_Focus::getVerse (request);
  
  fragments.push_back ("<div id=\"versepicker\">");
  
  fragments.push_back ("<input name=\"selectpassage\" id=\"selectpassage\" type=\"text\" value=\"" + bookName + " " + convert_to_string (chapter) + ":" + convert_to_string (verse) + "\" size=\"14\" />");
  fragments.push_back ("<input name=\"submitpassage\" id=\"submitpassage\" type=\"submit\" value=\"" + gettext("Go") + "\" />");
  
  fragments.push_back ("<div id=\"handpicker\">");
  
  fragments.push_back ("<select id=\"booklist\" size=15>");
  fragments.push_back (getBooksFragment (request, bible));
  fragments.push_back ("</select>");
  
  fragments.push_back ("<select id=\"chapterlist\" size=15>");
  fragments.push_back (getChaptersFragment (request, bible, book, chapter));
  fragments.push_back ("</select>");
  
  fragments.push_back ("<select id=\"verselist\" size=15 multiple>");
  fragments.push_back (getVersesFragment (request, bible, book, chapter, verse));
  
  fragments.push_back ("</select>");
  
  fragments.push_back ("</div>");
  
  fragments.push_back ("</div>");
  
  // The result.
  return filter_string_implode (fragments, "\n");
}


string Navigation_Passage::getBooksFragment (void * webserver_request, string bible)
{
  Webserver_Request * request = (Webserver_Request *) webserver_request;
  int activeBook = Ipc_Focus::getBook (request);
  // Take standard books in case of no Bible.
  vector <int> books;
  if (bible == "") {
    books = Database_Books::getIDs ();
  } else {
    books = filter_passage_get_ordered_books (bible);
  }
  string html = "";
  for (auto book : books) {
    string bookName = Database_Books::getEnglishFromId (book);
    string selected;
    if (book == activeBook) selected = " selected";
    html += "<option" + selected + ">" + bookName + "</option>";
  }
  return html;
}


string Navigation_Passage::getChaptersFragment (void * webserver_request, string bible, int book, int chapter)
{
  vector <int> chapters;
  if (bible == "") {
    Database_Versifications database_versifications = Database_Versifications ();
    chapters = database_versifications.getChapters ("English", book, true);
  } else {
    Webserver_Request * request = (Webserver_Request *) webserver_request;
    chapters = request->database_bibles()->getChapters (bible, book);
  }
  string html = "";
  for (auto ch : chapters) {
    string selected = "";
    if (ch == chapter) selected = " selected";
    html += "<option" + selected + ">" + convert_to_string (ch) + "</option>";
  }
  return html;
}


string Navigation_Passage::getVersesFragment (void * webserver_request, string bible, int book, int chapter, int verse)
{
  vector <int> verses;
  if (bible == "") {
    Database_Versifications database_versifications = Database_Versifications ();
    verses = database_versifications.getVerses ("English", book, chapter);
  } else {
    Webserver_Request * request = (Webserver_Request *) webserver_request;
    verses = usfm_get_verse_numbers (request->database_bibles()->getChapter (bible, book, chapter));
  }
  string html = "";
  for (auto vs : verses) {
    string selected = "";
    if (verse == vs) selected = " selected";
    html += "<option" + selected + ">" + convert_to_string (vs) +  "</option>";
  }
  return html;
}


string Navigation_Passage::code (string bible, bool header)
{
  string code = "";
  if (header) {
    code += "<script type=\"text/javascript\" src=\"/navigation/passage.js\"></script>";
    code += "\n";
  }
  code += "<script type=\"text/javascript\">";
  code += "\n";
  code += "navigationBible = \"" + bible + "\";";
  code += "\n";
  code += "</script>";
  code += "\n";
  return code;
}


void Navigation_Passage::setBookChapterVerse (void * webserver_request, int book, int chapter, int verse)
{
  Ipc_Focus::set (webserver_request, book, chapter, verse);
  recordHistory (webserver_request, book, chapter, verse);
}


void Navigation_Passage::setPassage (void * webserver_request, string bible, string passage)
{
  int currentBook = Ipc_Focus::getBook (webserver_request);
  int currentChapter = Ipc_Focus::getChapter (webserver_request);
  int currentVerse = Ipc_Focus::getVerse (webserver_request);
  passage = filter_string_trim (passage);
  Passage passage_to_set;
  if ((passage == "") || (passage == "+")) {
    passage_to_set = Navigation_Passage::getNextVerse (webserver_request, bible, currentBook, currentChapter, currentVerse);
  } else if (passage == "-") {
    passage_to_set = Navigation_Passage::getPreviousVerse (webserver_request, bible, currentBook, currentChapter, currentVerse);
  } else {
    Passage inputpassage = Passage ("", currentBook, currentChapter, convert_to_string (currentVerse));
    passage_to_set = filter_passage_interpret_passage (inputpassage, passage);
  }
  if (passage_to_set.book != 0) {
    Ipc_Focus::set (webserver_request, passage_to_set.book, passage_to_set.chapter, convert_to_int (passage_to_set.verse));
    Navigation_Passage::recordHistory (webserver_request, passage_to_set.book, passage_to_set.chapter, convert_to_int (passage_to_set.verse));
  }
}


Passage Navigation_Passage::getNextVerse (void * webserver_request, string bible, int book, int chapter, int verse)
{
  verse++;
  if (bible != "") {
    Webserver_Request * request = (Webserver_Request *) webserver_request;
    vector <int> verses = usfm_get_verse_numbers (request->database_bibles()->getChapter (bible, book, chapter));
    if (find (verses.begin(), verses.end(), verse) == verses.end()) {
      if (!verses.empty()) verse = verses.back ();
    }
  }
  Passage passage = Passage ("", book, chapter, convert_to_string (verse));
  return passage;
}


Passage Navigation_Passage::getPreviousVerse (void * webserver_request, string bible, int book, int chapter, int verse)
{
  verse--;
  if (bible != "") {
    Webserver_Request * request = (Webserver_Request *) webserver_request;
    vector <int> verses = usfm_get_verse_numbers (request->database_bibles()->getChapter (bible, book, chapter));
    if (find (verses.begin(), verses.end(), verse) == verses.end()) {
      if (!verses.empty ()) verse = verses [0];
    }
  }
  Passage passage = Passage ("", book, chapter, convert_to_string (verse));
  return passage;
}


void Navigation_Passage::gotoNextVerse (void * webserver_request, string bible)
{
  int currentBook = Ipc_Focus::getBook (webserver_request);
  int currentChapter = Ipc_Focus::getChapter (webserver_request);
  int currentVerse = Ipc_Focus::getVerse (webserver_request);
  Passage passage = Navigation_Passage::getNextVerse (webserver_request, bible, currentBook, currentChapter, currentVerse);
  if (passage.book != 0) {
    Ipc_Focus::set (webserver_request, passage.book, passage.chapter, convert_to_int (passage.verse));
    Navigation_Passage::recordHistory (webserver_request, passage.book, passage.chapter, convert_to_int (passage.verse));
  }
}


void Navigation_Passage::gotoPreviousVerse (void * webserver_request, string bible)
{
  int currentBook = Ipc_Focus::getBook (webserver_request);
  int currentChapter = Ipc_Focus::getChapter (webserver_request);
  int currentVerse = Ipc_Focus::getVerse (webserver_request);
  Passage passage = Navigation_Passage::getPreviousVerse (webserver_request, bible, currentBook, currentChapter, currentVerse);
  if (passage.book != 0) {
    Ipc_Focus::set (webserver_request, passage.book, passage.chapter, convert_to_int (passage.verse));
    Navigation_Passage::recordHistory (webserver_request, passage.book, passage.chapter, convert_to_int (passage.verse));
  }
}


void Navigation_Passage::recordHistory (void * webserver_request, int book, int chapter, int verse)
{
  Webserver_Request * request = (Webserver_Request *) webserver_request;
  string user = request->session_logic()->currentUser ();
  Database_Navigation database_navigation = Database_Navigation ();
  database_navigation.record (filter_string_date_seconds_since_epoch (), user, book, chapter, verse);
}


void Navigation_Passage::goBack (void * webserver_request)
{
  Webserver_Request * request = (Webserver_Request *) webserver_request;
  Database_Navigation database_navigation = Database_Navigation ();
  string user = request->session_logic()->currentUser ();
  Passage passage = database_navigation.getPrevious (user);
  if (passage.book) {
    Ipc_Focus::set (webserver_request, passage.book, passage.chapter, convert_to_int (passage.verse));
  }
}


void Navigation_Passage::goForward (void * webserver_request)
{
  Webserver_Request * request = (Webserver_Request *) webserver_request;
  Database_Navigation database_navigation = Database_Navigation ();
  string user = request->session_logic()->currentUser ();
  Passage passage = database_navigation.getNext (user);
  if (passage.book) {
    Ipc_Focus::set (webserver_request, passage.book, passage.chapter, convert_to_int (passage.verse));
  }
}
